src/coreclr/src/jit/hwintrinsic.cpp

```diff
@@ -658,9 +658,63 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
         baseType = getBaseTypeAndSizeOfSIMDType(sig->retTypeSigClass, &sizeBytes);
         retType  = getSIMDTypeForSize(sizeBytes);
         assert(sizeBytes != 0);
+
+        // Immediately return if this is non-arithmetic type.
+        if (!varTypeIsArithmetic(baseType))
```

Comment:
Why have the logic in both places, rather than waiting to check after the base type has been definitely determined?

Reply:
That was my initial thought, but there are cases where we want to return `nullptr` early rather than getting `baseType` from arg. E.g. `default(Vector256<ulong>).As<ulong, bool>()` should throw `NotSupportedException` but if we have logic later, we would end up having `baseType`  as `TYP_ULONG` and won't throw.

Reply:
A comment explaining that would be helpful for future refactoring. It might also be nice to consolidate the asserts as the few differences don't matter that much.

Reply:
I will add the comment.

>  It might also be nice to consolidate the asserts as the few differences don't matter that much.

Sorry, I didn't follow. Could you please elaborate?

Reply:
Maybe a helper method so you can do:
```cpp
if (!isSupportedBaseType(intrinsic, baseType))
{
    return nullptr;
}
```

where:
```cpp
static bool isSupportedBaseType(NamedIntrinsic intrinsic, var_types baseType) {
    // We don't actually check the intrinsic outside of the false case as we expect
    // the exposed managed signatures are either generic and support all types
    // or they are explicit and support the type indicated.

    if (varTypeIsArithmetic(baseType)) {
        return true;
    }

    assert((intrinsic == ...) || (intrinsic == ...) || ...);
    return false;
}
```

Reply:
> Sorry, I didn't follow. Could you please elaborate?

I was indicating that the fact that `AsByte` can't be hit in the `retType == TYP_STRUCT` scenario is probably not important and consolidating the two asserts together would likely be simpler.
The assert is really just to catch unexpected cases where we would need to debug anyways and so having it be exact for these two almost identical paths, rather than just sharing the logic, is probably unnecessary.